### PR TITLE
Feature/chol2inv

### LINKED ISSUE #648: Lack of function for computing the inverse of a matrix from its Cholesky factor
#### Summary:

We do not have a function that inputs a Cholesky factor `L` and outputs `(L * L')^(-1)`.

#### Description:

Such a function (for upper triangular Cholesky factors) exists in LAPACK and R under the name `chol2inv`.

#### Expected Output:

`(L')^(-1) * L^(-1)`

#### Additional Information:

I have written such a function.

#### Current Version:
v2.17.0

### PR DESCRIPTION
**Pull request by bgoodri**
#### Submission Checklist

- [X] Run unit tests: `./runTests.py test/unit`
- [X] Run cpplint: `make cpplint`
- [X] Declare copyright holder and open-source license: see below

#### Summary:

Introduce a function that inputs a matrix's Cholesky factor and outputs the matrix's inverse.

#### Intended Effect:

Fixes #648 

#### How to Verify:

Run the tests

#### Side Effects:

None

#### Documentation:

Will have to be added in a pull request for Stan 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Trustees of Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
I'm out until Tuesday---maybe @seantalts or @syclik can review.  Or I can get to it next week.

**Comment by bob-carpenter** [comment]
@bgoodri --- for future reference, it needs to be `[X]` instead of `[ X]` if you want check boxes to render properly.

File: stan/math/prim/mat/fun/chol2inv.hpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        return X;
+      }
+      matrix_t L_inv = mdivide_left_tri_low(L, matrix_t::Identity(K, K).eval());
+      matrix_t X(K, K);
+      for (int k = 0; k < K; ++k) {
+        X.coeffRef(k, k) = dot_self(L_inv.col(k).tail(K - k).eval());
+        for (int j = k + 1; j < K; ++j) {
+          int Kmj = K - j;
+          X.coeffRef(k, j) = X.coeffRef(j, k) = dot_product(
+            L_inv.col(k).tail(Kmj).eval(), L_inv.col(j).tail(Kmj).eval());
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/chol2inv.hpp]: Why is `.eval()` used here?  Is there an aliasing problem somewhere I don't see?

File: test/unit/math/prim/mat/fun/chol2inv_test.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
+#include <stan/math/prim/mat.hpp>
```
Comment (bob-carpenter) [inline test/unit/math/prim/mat/fun/chol2inv_test.cpp]: We want derivative tests for rev, fwd, and mix modes.  Here you have a second way to evaluate the function so you can just test that the derivatives are the same for a couple simple cases.

**Comment by bob-carpenter** [comment]
I can write these tests for you if you don't want to do it.   I'd add it along with a simple framework for testing two template functions had the same values and derivatives.  We have that for simple functions but not vector functions.

**Comment by bgoodri** [comment]
If you remove the `.eval()` calls, then the unit test does not compile:
```
In file included from ./stan/math/prim/mat.hpp:64:0,
                 from test/unit/math/prim/mat/fun/chol2inv_test.cpp:1:
./stan/math/prim/mat/fun/chol2inv.hpp: In instantiation of ‘Eigen::Matrix<LhsScalar, -1, -1, 0> stan::math::chol2inv(const Eigen::Matrix<LhsScalar, -1, -1, 0>&) [with T = double]’:
test/unit/math/prim/mat/fun/chol2inv_test.cpp:12:3:   required from here
./stan/math/prim/mat/fun/chol2inv.hpp:37:44: error: no matching function for call to ‘mdivide_left_tri_low(const Eigen::Matrix<double, -1, -1>&, const IdentityReturnType)’
       matrix_t L_inv = mdivide_left_tri_low(L, matrix_t::Identity(K, K));
                        ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ./stan/math/prim/mat/fun/chol2inv.hpp:7:0,
                 from ./stan/math/prim/mat.hpp:64,
                 from test/unit/math/prim/mat/fun/chol2inv_test.cpp:1:
./stan/math/prim/mat/fun/mdivide_left_tri_low.hpp:38:5: note: candidate: template<class T1, class T2, int R1, int C1, int R2, int C2> Eigen::Matrix<typename boost::math::tools::promote_args<RT1, RT2>::type, R1, C2> stan::math::mdivide_left_tri_low(const Eigen::Matrix<ViewElt, R, C>&, const Eigen::Matrix<T2, R2, C2>&)
     mdivide_left_tri_low(const Eigen::Matrix<T1, R1, C1> &A,
     ^~~~~~~~~~~~~~~~~~~~
./stan/math/prim/mat/fun/mdivide_left_tri_low.hpp:38:5: note:   template argument deduction/substitution failed:
In file included from ./stan/math/prim/mat.hpp:64:0,
                 from test/unit/math/prim/mat/fun/chol2inv_test.cpp:1:
./stan/math/prim/mat/fun/chol2inv.hpp:37:44: note:   ‘const IdentityReturnType {aka const Eigen::CwiseNullaryOp<Eigen::internal::scalar_identity_op<double>, Eigen::Matrix<double, -1, -1> >}’ is not derived from ‘const Eigen::Matrix<T2, R2, C2>’
       matrix_t L_inv = mdivide_left_tri_low(L, matrix_t::Identity(K, K));
                        ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ./stan/math/prim/mat/fun/chol2inv.hpp:7:0,
                 from ./stan/math/prim/mat.hpp:64,
                 from test/unit/math/prim/mat/fun/chol2inv_test.cpp:1:
./stan/math/prim/mat/fun/mdivide_left_tri_low.hpp:47:5: note: candidate: template<class T, int R1, int C1> Eigen::Matrix<T, R, C> stan::math::mdivide_left_tri_low(const Eigen::Matrix<T, R, C>&)
     mdivide_left_tri_low(const Eigen::Matrix<T, R1, C1> &A) {
     ^~~~~~~~~~~~~~~~~~~~
./stan/math/prim/mat/fun/mdivide_left_tri_low.hpp:47:5: note:   template argument deduction/substitution failed:
In file included from ./stan/math/prim/mat.hpp:64:0,
                 from test/unit/math/prim/mat/fun/chol2inv_test.cpp:1:
./stan/math/prim/mat/fun/chol2inv.hpp:37:44: note:   candidate expects 1 argument, 2 provided
       matrix_t L_inv = mdivide_left_tri_low(L, matrix_t::Identity(K, K));
                        ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./stan/math/prim/mat/fun/chol2inv.hpp:40:36: error: no matching function for call to ‘dot_self(Eigen::DenseBase<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true> >::SegmentReturnType)’
         X.coeffRef(k, k) = dot_self(L_inv.col(k).tail(K - k));
                            ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ./stan/math/prim/mat/fun/chol2inv.hpp:5:0,
                 from ./stan/math/prim/mat.hpp:64,
                 from test/unit/math/prim/mat/fun/chol2inv_test.cpp:1:
./stan/math/prim/mat/fun/dot_self.hpp:18:19: note: candidate: template<int R, int C> double stan::math::dot_self(const Eigen::Matrix<double, R, C>&)
     inline double dot_self(const Eigen::Matrix<double, R, C>& v) {
                   ^~~~~~~~
./stan/math/prim/mat/fun/dot_self.hpp:18:19: note:   template argument deduction/substitution failed:
In file included from ./stan/math/prim/mat.hpp:64:0,
                 from test/unit/math/prim/mat/fun/chol2inv_test.cpp:1:
./stan/math/prim/mat/fun/chol2inv.hpp:40:36: note:   ‘Eigen::DenseBase<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true> >::SegmentReturnType {aka Eigen::VectorBlock<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true>, -1>}’ is not derived from ‘const Eigen::Matrix<double, R, C>’
         X.coeffRef(k, k) = dot_self(L_inv.col(k).tail(K - k));
                            ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
./stan/math/prim/mat/fun/chol2inv.hpp:43:60: error: no matching function for call to ‘dot_product(Eigen::DenseBase<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true> >::SegmentReturnType, Eigen::DenseBase<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true> >::SegmentReturnType)’
           X.coeffRef(k, j) = X.coeffRef(j, k) = dot_product(
                                                 ~~~~~~~~~~~^
             L_inv.col(k).tail(Kmj), L_inv.col(j).tail(Kmj));
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
In file included from ./stan/math/prim/mat/fun/chol2inv.hpp:6:0,
                 from ./stan/math/prim/mat.hpp:64,
                 from test/unit/math/prim/mat/fun/chol2inv_test.cpp:1:
./stan/math/prim/mat/fun/dot_product.hpp:22:19: note: candidate: template<int R1, int C1, int R2, int C2> double stan::math::dot_product(const Eigen::Matrix<double, R, C>&, const Eigen::Matrix<double, R2, C2>&)
     inline double dot_product(const Eigen::Matrix<double, R1, C1>& v1,
                   ^~~~~~~~~~~
./stan/math/prim/mat/fun/dot_product.hpp:22:19: note:   template argument deduction/substitution failed:
In file included from ./stan/math/prim/mat.hpp:64:0,
                 from test/unit/math/prim/mat/fun/chol2inv_test.cpp:1:
./stan/math/prim/mat/fun/chol2inv.hpp:43:60: note:   ‘Eigen::DenseBase<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true> >::SegmentReturnType {aka Eigen::VectorBlock<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true>, -1>}’ is not derived from ‘const Eigen::Matrix<double, R, C>’
           X.coeffRef(k, j) = X.coeffRef(j, k) = dot_product(
                                                 ~~~~~~~~~~~^
             L_inv.col(k).tail(Kmj), L_inv.col(j).tail(Kmj));
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
In file included from ./stan/math/prim/mat/fun/chol2inv.hpp:6:0,
                 from ./stan/math/prim/mat.hpp:64,
                 from test/unit/math/prim/mat/fun/chol2inv_test.cpp:1:
./stan/math/prim/mat/fun/dot_product.hpp:35:19: note: candidate: double stan::math::dot_product(const double*, const double*, size_t)
     inline double dot_product(const double* v1, const double* v2,
                   ^~~~~~~~~~~
./stan/math/prim/mat/fun/dot_product.hpp:35:19: note:   candidate expects 3 arguments, 2 provided
./stan/math/prim/mat/fun/dot_product.hpp:48:19: note: candidate: double stan::math::dot_product(const std::vector<double>&, const std::vector<double>&)
     inline double dot_product(const std::vector<double>& v1,
                   ^~~~~~~~~~~
./stan/math/prim/mat/fun/dot_product.hpp:48:19: note:   no known conversion for argument 1 from ‘Eigen::DenseBase<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true> >::SegmentReturnType {aka Eigen::VectorBlock<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true>, -1>}’ to ‘const std::vector<double>&’
<builtin>: recipe for target 'test/unit/math/prim/mat/fun/chol2inv_test.o' failed
make: *** [test/unit/math/prim/mat/fun/chol2inv_test.o] Error 1
make -j1 test/unit/math/prim/mat/fun/chol2inv_test failed
```

**Comment by bgoodri** [comment]
I suppose it would be good to have a function that compares derivatives with another function. But if the derivatives of `mdivide_left_tri_low`, `dot_self`, and / or `dot_product` are correct and the autodiff is correct, then `chol2inv` is going to be fine.

**Comment by bob-carpenter** [comment]
Thanks for the clarification.  I forgot that our functions weren't properly generalized to deal with expression template arguments.  

As to testing, I'll try to write the tests for this.  It's the next thing on my plate. 

I'm OK merging this with or without autodiff tests, so I'm going to go ahead and approve, but I think we should ask @syclik about this.

**Comment by bgoodri** [comment]
For something like this, one could start with L, compose Sigma = L * L',
invert Sigma, and see what the derivatives of each element in the lower
triangle of the precision matrix are with respect to each lower triangular
element of L. But you lose a lot of numerical accuracy along with way,
which is why a function like chol2inv needs to exist so that you can go
from L directly to the precision matrix. So, unless the tolerance is pretty
loose, the check wouldn't pass. And if the tolerance is loose, I don't know
that it tells us very much.

**Comment by bob-carpenter** [comment]
Understood.  I've had no problems with simpler functions and tolerances of 1e-7 or so, but this could be a lot more problematic.  

Here, we have two functions, a simple one and the compound one, so we can test analytic vs. analytic.  I just want to add a framework to do that more easily so each one of these functions isn't a pain to test.

Probably don't want to wait with this pull request on this as I have a bunch of other stuff stacked up.

**Comment by syclik** [comment]
@bob-carpenter, feel free to merge.

If it's easy to create a quick test case, I'd say put up a simple unit test with something that's easy to verify, like a special case. Or have loose tolerances on something that's a little harder. @bgoodri: what it tells us and what it prevents are gross errors.

**Comment by bob-carpenter** [comment]
OK, I'll add a quick test case and/or merge ASAP.  I want to think about these tests, so I don't mind doing it.

**Comment by syclik** [comment]
Thanks, Bob!

**Comment by bgoodri** [comment]
If you were going to test stuff like this, then use the fact that

I = t(L) * precision * L

is an identity matrix and any norm of the Jacobian should be zero.

But that doesn't test chol2inv; that tests the functions that chol2inv
calls, which are already being tested narrowly.
If we can't trust that compositions of functions work correctly, then we
can't have people using Stan.

**Comment by sakrejda** [comment]
@bgoodri isn't the whole point of numerical computation that compositions of functions, in general, do not work?

**Comment by bgoodri** [comment]
They don't work great, but there is no way way there can be a gross error
in the derivatives of chol2inv without there being a gross error in the
derivatives of one of the functions it calls or in the autodiff mechanism.

Apart from the unit tests not testing some things, the biggest problem is
the amount of redundancy that is built into the unit tests. And this is a
(small) example of that.

**Comment by sakrejda** [comment]
Would be great if we could be coherent about deciding where to place the testing.

**Comment by bob-carpenter** [comment]
> ...
> But that doesn't test chol2inv; that tests the functions that chol2inv
> calls, which are already being tested narrowly.

Right---no need to do this kind of test.

> If we can't trust that compositions of functions work correctly, then we
> can't have people using Stan.

Agreed.

I don't trust our developers to compose functions correctly to implement a target mathematical function.  I've messed up things as simple as chol2inv() myself.  

One could argue that testing the values and inspecting that it only uses simple tested Stan functions, which is why I thought this one would be OK without tests.

Having the tests help future-proof vs. someone adding analytic derivatives.  There I do want to test thoroughly in all combinations.

**Comment by bob-carpenter** [comment]
Tests for functions like this are fast.  If we can add a framework to do this everywhere, the test code will be simple and thorough.  Remains for me to show that, though.

I don't think things like this are too compute intensive, either.  The distributions are the killers, and those aren't simple composed functions.

**Comment by bgoodri** [comment]
It would make sense to have a utility called test_identity() or something
that takes a lambda function and a vector and verifies both the levels and
the norm of the Jacobian. But that would mainly be for developer
convenience.

**Comment by bob-carpenter** [comment]
Right.  That's the kind of thing I'm thinking about for a generalized framework.  No way we can sustain writing four files full of specific tests with our rather clunky autodiff framework.

**Comment by bob-carpenter** [comment]
Thanks, @bgoodri.  I looked through what I need for a testing framework and it's going to take more than a couple hours, but should be doable.  So I just went ahead and merged this one.  I'll swing back through with the higher-order autodiff tests as I get to them.

**Pull request merged**